include/net/net_pkt.h

```diff
@@ -526,6 +557,19 @@ static inline void net_pkt_set_ipv6_hop_limit(struct net_pkt *pkt,
 }
 #endif /* CONFIG_NET_IPV6 */
 
+static inline u16_t net_pkt_ip_opts_len(struct net_pkt *pkt)
+{
+#if defined(CONFIG_NET_IPV6)
+	return pkt->ipv6_ext_len;
+#elif defined(CONFIG_NET_IPV4)
+	return pkt->ipv4_opts_len;
```

Comment:
This and the union where the value is stored is a bit unnecessary. This function always returns u16_t, so one u16_t can hold both IPv4 options length and IPv6 extension header length. The code requires IPv4 options length to be u8_t when setting it with net_pkt_set_ipv4_opts_len() , so the value to set will be fine. And as the value returned is always u16_t, the u8_t pkt->ipv4_opts_len is there only to save 8 bits if IPv6 is not used. Is the union strictly necessary, could this just return the one and only u16_t, or do we save 1 byte in the struct when IPv6 is not used?

Right now two consecutive returns look confusing, until one notices that it is a union.

Reply:
>  or do we save 1 byte in the struct when IPv6 is not used?

That's the idea yes.

Imo the union and these ifdefs are fine. It is not super clean but I can't see other way to save 1 byte if ipv6 is not used. 

Unless we don't care about saving 1 byte when possible, but we have been doing this mistake before: once we drop our attention on being strict on memory consumption, it will get bloated little by little

Reply:
That's a good point. Let's move on.